src/models/job.cr

```diff
@@ -5,15 +5,39 @@ module Mssh
       commands: Array(String),
     )
 
-    def queued(groups : Groups)
+    def process_commands(variables : Variables?)
+      @commands.each_with_index do |cmd, i|
```

Comment:
These implementations are little bit unstable. How about using regex? You could simply write the below by `gsub`.
Refer https://crystal-lang.org/api/0.27.2/String.html#gsub%28pattern%3ARegex%2Chash%3AHash%28String%2CUNDERSCORE%29%7CNamedTuple%29-instance-method

Reply:
For example
```crystal
variables = [
  {name: "A", value: "AA"},
  {name: "B", value: "BB"}
]

commands = [
  "echo ${A} && echo ${B}",
  "echo ${A}"
]

commands.each_with_index do |cmd, i|
  commands[i] = variables.reduce(cmd) { |c, v| c.gsub(/\${#{v[:name]}}/, v[:value]) }
end

puts commands
# => ["echo AA && echo BB", "echo AA"]
```

Reply:
Sample code with nullable variables
```crystal
#
# Replace `Array(NamedTuple(name: String, value: String))` into `Variables?`!
#
def process_commands(variables : Array(NamedTuple(name: String, value: String))?)
  commands = [
    "echo ${A} && echo ${B}",
    "echo ${A}"
  ]

  return commands unless variables_not_nil = variables

  commands.each_with_index do |cmd, i|
    commands[i] = variables_not_nil.reduce(cmd) { |c, v| c.gsub(/\${#{v[:name]}}/, v[:value]) }
  end

  puts commands
end

process_commands([{name: "A", value: "AA"}, {name: "B", value: "BB"}])
process_commands(nil)
```

Reply:
Thanks for explanation. It's so cool 👌